### ISSUE TITLE
Entering a previously visited composite state re-enters its last active substate even though no history states are present

### ISSUE THREAD

**Issue by Ulenspiegel** [boost-experimental/msm-lite#55]
It appears that after leaving and re-entering a composite state, it enters the substate it was in when it was exited, not the initial state, as though there's history stored (though there's no history specified for the composite):

```c++
#include <cstdio>
#include <boost/msm-lite.hpp>

namespace msm = boost::msm::lite;
using msm::state;
using msm::sm;

struct e1 {};
struct e2 {};
struct e3 {};
struct e4 {};

struct SubState {
  auto configure() const noexcept {
    using namespace msm;
    using msm::on_exit;
    state<class A> a;
    state<class B> b;
    state<class C> c;
    return make_transition_table(
        *a + event<e1> = b,
        b + event<e2> = c,
        a + on_entry / [] { printf("Entering s(a)\n"); },
        b + on_entry / [] { printf("Entering s(b)\n"); },
        c + on_entry / [] { printf("Entering s(c)\n"); },
        a + on_exit / [] { printf("Leaving s(a)\n"); },
        b + on_exit / [] { printf("Leaving s(b)\n"); },
        c + on_exit / [] { printf("Leaving s(c)\n"); }
    );
  }
};

struct TopState {
  auto configure() const noexcept {
    using namespace msm;
    using msm::on_exit;
    state<sm<SubState>> s;
    state<class X> x;
    return make_transition_table(
        *"init1"_s = s,
        s + event<e3> = x,
        x + event<e4> = s,
        s + on_entry / [] { printf("Entering s\n"); },
        x + on_entry / [] { printf("Entering x\n"); },
        s + on_exit / [] { printf("Leaving s\n"); },
        x + on_exit / [] { printf("Leaving x\n"); }
    );
  }
};

int main(int argc, char** argv) {
  using namespace msm;
  sm<TopState> m;
  printf("\nProcessing e1.\n");
  m.process_event(e1{});
  printf("\nProcessing e2.\n");
  m.process_event(e2{});
  printf("\nProcessing e3.\n");
  m.process_event(e3{});
  printf("\nProcessing e4.\n");
  m.process_event(e4{});
  return 0;
}
```

It ouputs the following:

```
Entering s(a)
Entering s

Processing e1.
Leaving s(a)
Entering s(b)

Processing e2.
Leaving s(b)
Entering s(c)

Processing e3.
Leaving s(c)
Leaving s
Entering x

Processing e4.
Leaving x
Entering s(c)
Entering s
```

Re-entering composite state `s` activated its substate `c` (last activated before transition `s -> x`), though it should have entered its initial substate.

**Comment by krzysztof-jusiak** [comment]
Thank you @Ulenspiegel, you are absolutely awesome!
I'm tracking all issues you are reporting and can't wait to fix them all.
I'm just a bit busy at the moment with the release of a new version of experimental boost.di.
As soon as that will be done I'm dying to get into msm-lite and tackle all these issues.
Thanks once again, I really do appreciate your work and issues reported.

**Comment by Ulenspiegel** [comment]
There's no pressure! Most of that can be worked around with a bit of extra effort. And I'm glad to do the testing; a lot of that comes from battlefield, and I'm having good time baptizing the library in an actual project (which uses the experimental boost.di too, speaking of that). Don't feel pressed, and thanks again for great work. :)

**Comment by krzysztof-jusiak** [comment]
Fixed by eb0b9bfa2120225da52ba828a9022aa4c0ff7a19
